src/Iterator/TypedIterator.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+/**
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace loophp\collection\Iterator;
+
+use Generator;
+use InvalidArgumentException;
+
+use function count;
+use function gettype;
+use function in_array;
+
+/**
+ * @internal
+ *
+ * @template TKey
+ * @template T
+ *
+ * @extends ProxyIterator<TKey, T>
+ */
+final class TypedIterator extends ProxyIterator
+{
+    /**
+     * @param iterable<TKey, T> $iterable
+     */
+    public function __construct(iterable $iterable)
+    {
+        $this->iterator = new ClosureIterator(
+            static function (iterable $iterator): Generator {
+                $allowedTypes = ['NULL'];
+
+                foreach ($iterator as $key => $value) {
+                    $currentType = gettype($value);
+
+                    if (count($allowedTypes) < 2) {
+                        $allowedTypes[] = $currentType;
+                    }
+
+                    if (!in_array($currentType, $allowedTypes, true)) {
+                        throw new InvalidArgumentException(
+                            sprintf(
+                                'Detected mixed types: %s and %s!',
+                                $allowedTypes[1],
+                                $currentType
+                            )
+                        );
+                    }
+
+                    yield $key => $value;
+                }
+            },
+            $iterable
+        );
+    }
+}
```

Comment:
@drupol added this now based on Idea 1, with a few minor modifications. Please see the [spec](https://github.com/loophp/collection/pull/102/files#diff-34c434dbd5917f3fc5a051c137420d165761f1eef3805d8ba9abbbc49fbbf9b2R18) to understand the complete behaviour.

DONE:
- [x] specs for all scalars: `int, float, string, bool`
- [x] allow `NULL` in addition to another type: I thought this is best because having nullable types is quite common in php; I think it would be too restrictive to _not_ allow it

TODO:
- [ ] implement logic and tests for other types: `resource, array, object`. For object check class, for resource check to not have a mix of closed and not closed, and for array I think don't check inside for now
- [ ] use `TypedIterator` in the `strict` operation like I described [here](https://github.com/loophp/collection/pull/102#discussion_r646189027)

Reply:
Hi!

This morning I updated it, let me know what you think: https://gist.github.com/drupol/15ddab9628e4cd2f90321ad2f0368bd9

Reply:
Morning, that's interesting so you want to allow user the user the option to define how the typing is enforced. Sounds good to me :).

Another point:
- I think `yield` should happen after the type checking. If you try to run the tests with your version you will see that they only pass on the call to `next` after you passed the element which was of a different type

I will use the gist to update this when I get the chance and add more tests, thanks!

Reply:
> Morning, that's interesting so you want to allow user the user the option to define how the typing is enforced. Sounds good to me :).

Yes, there is a reason to that.
This `getType` is a kind of basic implementation, it will never cover all the user needs. Then, in those specific case, the user is able to customize that `callable`.
If you checkout the operations, I often let the user customize the stuff with callables, so it's fully flexible. I'm trying to never define/set something in stone within an operation.

> Another point:
> 
>     * I think `yield` should happen after the type checking. If you try to run the tests with your version you will see that they only pass on the call to `next` after you passed the element which was of a different type

Mmmh but this is what I did, the `yield` statement is after the type checks.

> I will use the gist to update this when I get the chance and add more tests, thanks!

Sure, let's see what comes out.

Reply:
> Mmmh but this is what I did, the yield statement is after the type checks.

Oh you're right, it only yields early if the value is `null` 👍